prometheus/metrics.go

```diff
@@ -152,71 +152,86 @@ func getServiceHealth(api v1.API, namespace, servicename string, ports []int32)
 }
 
 func getMetrics(api v1.API, q *MetricsQuery) Metrics {
-	labelsIn, labelsOut, labelsErrorIn, labelsErrorOut := buildLabelStrings(q)
+	labelsIn, labelsErrorIn := buildLabelStrings(q, true)
 	groupingIn := strings.Join(q.ByLabelsIn, ",")
+	inboundMetrics := fetchAllMetrics(api, q, labelsIn, labelsErrorIn, groupingIn, "_in")
+	if q.Service != "" {
+		// If a service is set, stop now; we'll only have inbound metrics
+		return inboundMetrics
+	}
+
+	labelsOut, labelsErrorOut := buildLabelStrings(q, false)
 	groupingOut := strings.Join(q.ByLabelsOut, ",")
+	metrics := fetchAllMetrics(api, q, labelsOut, labelsErrorOut, groupingOut, "_out")
 
-	return fetchAllMetrics(api, q, labelsIn, labelsOut, labelsErrorIn, labelsErrorOut, groupingIn, groupingOut)
+	// Merge in a single object
+	for key, obj := range inboundMetrics.Metrics {
+		metrics.Metrics[key] = obj
+	}
+	for key, obj := range inboundMetrics.Histograms {
+		metrics.Histograms[key] = obj
+	}
+	return metrics
 }
 
-func buildLabelStrings(q *MetricsQuery) (string, string, string, string) {
-	labelsIn := []string{`reporter="destination"`}
-	labelsOut := []string{`reporter="source"`}
-	if config.Get().IstioNamespace == q.Namespace {
-		labelsOut = []string{`reporter="destination"`}
+func buildLabelStrings(q *MetricsQuery, isInbound bool) (string, string) {
+	var referential string
+	var labels []string
+	if isInbound {
+		referential = "destination"
+		labels = []string{`reporter="destination"`}
```

Comment:
I think it may be a problem using destination telemetry for service telemetry.  Only source telemetry will include the source-side failures.  I think we should be using destination telemetry only in two scenarios:

1.  We want to limit the data to only the server-side processing (like response-time of the service processing minus the network latency)
2. There is no source-side telemetry (the istio-system use-case)

I'm not sure we actually have a need for case 1 at the moment, so it seems to me that source-side telemetry should be the default in the metrics API.  I know this is not what has been done so far, but I'm not sure why.

Reply:
@jshaughn that makes sense, ok. I'm going to change it for all metrics.

Reply:
@jshaughn I just created a task for that: https://issues.jboss.org/browse/KIALI-1344
But that made me wonder if source-reporting telemetry really catches everything and in particular inbound telemetry when the inbound is not istio-proxied ? Such as "Unknown" workloads? That would be a problem.

Reply:
@jotak , I commented on the new Jira as well, but you are right and it is a problem.  Specifically, "unknown" source is only destination-side reporting.  I'm not sure exactly how we want to handle this, perhaps we should meet.  If we had to do one or the other (for now at least) I think  ensuring we report the source-side failure telemetry is perhaps most important.